src/librustc_mir/transform/split_local_fields.rs

```diff
@@ -0,0 +1,429 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::hir;
+use rustc::ty::{self, TyCtxt, Ty};
+use rustc::ty::util::IntTypeExt;
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::*;
+use rustc::mir::visit::{PlaceContext, MutVisitor, Visitor};
+use rustc::session::config::FullDebugInfo;
+use rustc_const_math::ConstInt;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use std::collections::BTreeMap;
+use syntax_pos::Span;
+use transform::{MirPass, MirSource};
+
+pub struct SplitLocalFields;
+
+impl MirPass for SplitLocalFields {
+    fn run_pass<'a, 'tcx>(&self,
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          source: MirSource,
+                          mir: &mut Mir<'tcx>) {
+        // Don't run on constant MIR, because trans might not be able to
+        // evaluate the modified MIR.
+        // FIXME(eddyb) Remove check after miri is merged.
+        let id = tcx.hir.as_local_node_id(source.def_id).unwrap();
+        match (tcx.hir.body_owner_kind(id), source.promoted) {
+            (_, Some(_)) |
+            (hir::BodyOwnerKind::Const, _) |
+            (hir::BodyOwnerKind::Static(_), _) => return,
+
+            (hir::BodyOwnerKind::Fn, _) => {
+                if tcx.is_const_fn(source.def_id) {
+                    // Don't run on const functions, as, again, trans might not be able to evaluate
+                    // the optimized IR.
+                    return
+                }
+            }
+        }
+
+        let mut collector = LocalPathCollector {
+            locals: mir.local_decls.iter().map(|decl| {
+                LocalPath::new(decl.ty)
+            }).collect()
+        };
+
+        // Can't split return and arguments.
+        collector.locals[RETURN_PLACE].make_opaque();
+        for arg in mir.args_iter() {
+            collector.locals[arg].make_opaque();
+        }
+
+        // We need to keep user variables intact for debuginfo.
+        if tcx.sess.opts.debuginfo == FullDebugInfo {
```

Comment:
I *really* like gating optimizations on debuginfo because then we get to optimize all the temporaries that the user can never see *anyway*, at a potentially acceptable cost for debug builds. IMO that'd be part of the cure to the "10x slower debug build" syndrome.

Reply:
> this is a "scalar replacement of aggregates" pass, that unrolls a single user variable into multiple ones. Any thoughts on how to handle it

DWARF already has the necessary machinery for this, and both gdb and lldb implement this.  LLVM can emit it as well. See, e.g., `test/DebugInfo/X86/split-global.ll` for an example.  In that test there is a `DIExpression` that uses an LLVM-only DWARF operator to describe this to LLVM (see `docs/LangRef.rst` for some more info on that)

Reply:
@tromey I think what I meant is: any thoughts on what it would take for *us* to emit the right thing? What would we have to add to MIR to do that?

@eddyb I think I'd be happy to open an issue to discuss this topic and tag that `if` with a FIXME.

Reply:
Hang on, I thought you could have overlapping variables, but not split ones. If both work I might try to implement them in this PR. cc @dotdash Maybe we can have debuginfo for scalar pairs without placing them in memory?

Reply:
See https://llvm.org/docs/LangRef.html#diexpression and https://llvm.org/docs/SourceLevelDebugging.html#llvm-dbg-value.  The idea is to emit an `llvm.dbg.value` that has a location expression describing how the scalar replacement was done.  In my tree I also see that `llvm::DIBuilder::insertDeclare` has an overload that accepts a location expression; I'm not sure how to reconcile that with the text saying that `llvm.dbg.declare` is deprecated.  This `DIBuilder` method is what LLVM's own SROA pass uses.

Reply:
Oh, sorry, if the question is about MIR and not LLVM, then I have no clue.  I haven't learned anything about MIR yet.

Reply:
We have to add support to MIR for this, and the LLVM information is what we were missing - we already emit expressions for closure upvars so we could generalize that.